### PR TITLE
RSDK-4844 Add better boilerplate for module development

### PR DESCRIPTION
**Pull request by benjirewis**
[RSDK-4844](https://viam.atlassian.net/browse/RSDK-4844)

- Renames `ModuleService_` to `ModuleService`
- Adds a `std::shared_ptr<Server> server_` field to `ModuleService`
   -  `ModuleService::server_` is created (not started) in the constructor
   -  `ModuleService::server_` is shutdown in the `ModuleService` destructor
 - Removes `std::shared_ptr<Server>` param from `add_model_from_registry` and `add_api_from_registry` 
 - Adds a `SignalManager signal_manager_` field to `ModuleService`
 - Changes the `ModuleService::start(std::shared_ptr<Server>)` method to `ModuleService::serve()` and alters its behavior
   - `serve` runs the `ModuleService::server_` until the `ModuleService::signal_manager_` finishes waiting
   - `serve` logs API/model pairs the module handles
 - Creates a new "boilerplate" overloaded `ModuleService` constructor `ModuleService(int argc, char** argv, std::vector<std::shared_ptr<ModelRegistration>>)`
   - Validates the socket path is present in `argv` and sets it on the class 
   - Sets global boost logger severity based on presence of the `--log-level` flag in `argv`
   - Registers the passed in `ModelRegistration` classes' models (not APIs)
   - Adds models and custom APIs from registry to `ModuleService`
 - Adds doc comments to `SignalManager` and `ModuleService` and makes more methods private

[RSDK-4844]: https://viam.atlassian.net/browse/RSDK-4844?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/viam/examples/modules/simple/main.cpp
Comment (benjirewis) [inline src/viam/examples/modules/simple/main.cpp:24]: Unnecessary.

File: src/viam/examples/modules/simple/main.cpp
Comment (benjirewis) [inline src/viam/examples/modules/simple/main.cpp:99]: Why can I not just pass {mr} below or use `auto` here? I don't know 🤷🏻 ; something something implicit conversion.

File: src/viam/examples/modules/tflite/main.cpp
Comment (benjirewis) [inline src/viam/examples/modules/tflite/main.cpp:746]: I've left the tflite module as an example of using the "old" constructor (`ModuleService(std::string socket_path)`), and doing arg checking + model checking manually. Signal handling and manual calls to the underlying `Server` should no longer be necessary.

File: src/viam/sdk/module/service.cpp
Comment (benjirewis) [inline src/viam/sdk/module/service.cpp:47]: I'm not sure why this was in an unnamed namespace; happy to keep it that way if desired, but I moved it to be a private method on `ModuleService`.

File: src/viam/sdk/module/service.cpp
Comment (benjirewis) [inline src/viam/sdk/module/service.cpp:261]: [RSDK-5509](https://viam.atlassian.net/browse/RSDK-5509) and an upcoming PR.

File: src/viam/sdk/module/service.hpp
Comment (benjirewis) [inline src/viam/sdk/module/service.hpp:94]: These could probably be `unique_ptr`s; I don't really know what I'm doing 😆 .

File: src/viam/sdk/module/service.cpp
Comment (benjirewis) [inline src/viam/sdk/module/service.cpp:253]: These logs look like:
```
[2023-12-01 16:26:38.114981] [0x00007f78e65fdd00] [info]    Module listening on /tmp/viam-module-1887335542/MyModule.sock
[2023-12-01 16:26:38.114981] [0x00007f78e65fdd00] [info]    Module handles the following API/model pairs:
API: viam:service:summation
    Model: viam:summation:mysummation
API: viam:component:gizmo
    Model: viam:gizmo:mygizmo
API: rdk:component:base
    Model: viam:base:mybase
```
Let me know if you prefer a different format or don't like the idea of this log generally.

File: src/viam/examples/modules/simple/main.cpp
Comment (stuqdog) [reply src/viam/examples/modules/simple/main.cpp:98]: huh, that surprises me.

File: src/viam/sdk/module/handler_map.cpp
Comment (stuqdog) [inline src/viam/sdk/module/handler_map.cpp:76]: (minor) per clang-tidy PR, can we replace `std::endl` with `"\n"`?

File: src/viam/sdk/module/service.hpp
Comment (stuqdog) [inline src/viam/sdk/module/service.hpp:3]: Can we move this include to be more consistent with our include sorting elsewhere?

File: src/viam/sdk/module/service.cpp
Comment (stuqdog) [reply src/viam/sdk/module/service.cpp:48]: I'm sure there was a reason but I can't recall what it is. Having it as a private method seems much better.

**Review by stuqdog — CHANGES_REQUESTED** [review]
Nice! A couple small formatting things but this looks great.

**Review by stuqdog — APPROVED** [review]
lgtm!

File: src/viam/examples/modules/simple/main.cpp
Comment (acmorrow) [inline src/viam/examples/modules/simple/main.cpp:57]: I think the better change is to say `auto&`. The `const`-ness will be inferred. Plain `auto` doesn't capture references.

https://godbolt.org/z/4h9b5xq9a

File: src/viam/sdk/module/service.cpp
Comment (acmorrow) [inline src/viam/sdk/module/service.cpp:232]: `for (const auto& mr :` or `for (auto&& mr :`

File: src/viam/examples/modules/simple/main.cpp
Comment (acmorrow) [reply src/viam/examples/modules/simple/main.cpp:98]: I'd expect both to work?

https://godbolt.org/z/5sPWjjjzx

What error do you get when you try?

File: src/viam/sdk/module/service.cpp
Comment (acmorrow) [reply src/viam/sdk/module/service.cpp:48]: Well, technically, it doesn't need any access to private state of the `ModuleService`, so it doesn't strictly need to be a member. In general, if something doesn't need to be a member of a class, it shouldn't be. I think this is basically an "implementation" helper needed only in this file to support defining `[Add|Reconfigure]Resource`. By making it a private member of `ModuleService`, it now has linkage/visibility even though it doesn't need it.

**Review by acmorrow — CHANGES_REQUESTED** [review]
First pass, these are small things and comments on existing discussions. Once these are cleaned up I want to take a closer look at the design and API of ModuleService. I suspect it can be improved a bit, but I don't want to try to both address the small stuff and redesign the API in the same pass.

File: src/viam/examples/modules/simple/main.cpp
Comment (benjirewis) [reply src/viam/examples/modules/simple/main.cpp:98]: I get: `No matching function for call to 'make_shared' (clang  ovl_no_viable_function_in_call)`

File: src/viam/sdk/module/service.cpp
Comment (benjirewis) [reply src/viam/sdk/module/service.cpp:48]: > Well, technically, it doesn't need any access to private state of the ModuleService, so it doesn't strictly need to be a member.

I think I was looking at `get_dependencies` and noticed it took a `ModuleService_* m`; that _looks_ a lot like a method on `ModuleService_` and not necessarily a helper function. I do see what you're saying about it not accessing private state, though. Happy to change back if you feel strongly.

File: src/viam/sdk/module/service.cpp
Comment (benjirewis) [reply src/viam/sdk/module/service.cpp:232]: Sounds good; used the first. Just curious, it seems like you prefer `auto` when type annotations are too long? Or is it when they're obvious?

File: src/viam/sdk/module/service.cpp
Comment (benjirewis) [reply src/viam/sdk/module/service.cpp:253]: They are indeed multiline and look a little gross when output by the RDK. NetCode is currently working on making module logging better, so I think I'd rather handle this at the RDK level instead of here (module authors should be able to log with `\n` and still get something vaguely greppable IMO).

File: src/viam/sdk/module/service.hpp
Comment (benjirewis) [reply src/viam/sdk/module/service.hpp:22]: Ah good point; moved to its own file `module/signal_manager.*pp`. I think I'd like to still expose it as a public class.

File: src/viam/sdk/module/service.hpp
Comment (benjirewis) [reply src/viam/sdk/module/service.hpp:29]: Cool; sounds good. I did replace `@param` with `@throws` to describe that it may throw. Not sure if that was necessary.

File: src/viam/sdk/module/service.hpp
Comment (benjirewis) [reply src/viam/sdk/module/service.hpp:42]: 👍🏻 added for all constructors.

File: src/viam/sdk/module/service.hpp
Comment (benjirewis) [reply src/viam/sdk/module/service.hpp:61]: Yeah I think we should derive privately now. Thanks for the cool example! I hadn't seen the godbolt playground before it's cool.

File: src/viam/sdk/module/service.hpp
Comment (benjirewis) [reply src/viam/sdk/module/service.hpp:94]: > Seems like signal_manager_ could just be held by value? Why does it need to be in a separate allocation?

It can be and done!

> I'm also not sure why Server can't be a unique_ptr, ditto Module.

Ah `module_` can be a unique_ptr (I thought it couldn't for some reason); changed. As for `server_`, from my other comment:

File: src/viam/examples/modules/simple/main.cpp
Comment (acmorrow) [reply src/viam/examples/modules/simple/main.cpp:98]: Hmm. Well, I don't think it is important enough to spend a lot of time on, the code you have is fine.

File: src/viam/sdk/module/service.cpp
Comment (acmorrow) [reply src/viam/sdk/module/service.cpp:48]: No, I don't feel strongly, its fine as a private member.

File: src/viam/sdk/module/service.cpp
Comment (acmorrow) [reply src/viam/sdk/module/service.cpp:247]: Alright, sounds good.

**Review by acmorrow — APPROVED** [review]
LGTM. I expect there will be further evolution of the ModuleService API, and the SignalManager too, but this is a meaningful step forward in SDK usability for module authors, so I think we should lock in the gains.

File: src/viam/sdk/module/service.cpp
Comment (acmorrow) [reply src/viam/sdk/module/service.cpp]: I'm fairly pro-`auto` in general. In this case the type is obvious. Some interesting reading here: https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/.

File: src/viam/examples/modules/simple/main.cpp
Comment (acmorrow) [reply src/viam/examples/modules/simple/main.cpp]: That's silly of the linter. It already is `const`. But I don't mind if you want to keep the `const` to placate the linter. Still, maybe we should consider turning that off: https://clang.llvm.org/extra/clang-tidy/checks/readability/qualified-auto.html#cmdoption-arg-AddConstToQualified.

Comment (benjirewis) [reply src/viam/examples/modules/simple/main.cpp]: > maybe we should consider turning that off

I'm not sure CI/our clang-tidy would complain. It might just be my local clang-tidy/IDE. I've left it without `const`, as it does seem redundant.

File: src/viam/sdk/module/service.cpp
Comment (benjirewis) [reply src/viam/sdk/module/service.cpp]: Ah interesting stuff, thanks for the reading material. My intuition was that `auto` was like "cheating" in C++ and it was usually better to declare type. That article was insightful.

**Comment by benjirewis** [comment]
This change breaks modules so will wait until Monday morning to merge and notify C++ module authors of breaking change.

**Pull request merged**
